lib/FreeRTOS/mqtt-agent-interface/freertos_command_pool.c

```diff
@@ -0,0 +1,142 @@
+ * as PUBLISH or SUBSCRIBE) between the command being created by an API call and
+ * completion of the command by the execution of the command's callback.
+ */
+static Command_t commandStructurePool[ MQTT_COMMAND_CONTEXTS_POOL_SIZE ];
+
+/**
+ * @brief A queue used to guard the pool of Command_t structures. Structures
+ * may be obtained by receiving a pointer from the queue, and returned by
+ * sending the pointer back into it.
+ */
+static AgentMessageContext_t commandStructQueue;
+
+/**
+ * @brief Initialization status of the queue.
+ */
+static volatile uint8_t initStatus = QUEUE_NOT_INITIALIZED;
+
+/*-----------------------------------------------------------*/
+
+void Agent_InitializePool( void )
+{
+    size_t i;
+    Command_t * pCommand;
+    static uint8_t staticQueueStorageArea[ MQTT_COMMAND_CONTEXTS_POOL_SIZE * sizeof( Command_t * ) ];
+    static StaticQueue_t staticQueueStructure;
+    bool commandAdded = false;
+
+    if( initStatus == QUEUE_NOT_INITIALIZED )
+    {
+        memset( ( void * ) commandStructurePool, 0x00, sizeof( commandStructurePool ) );
+        commandStructQueue.queue = xQueueCreateStatic( MQTT_COMMAND_CONTEXTS_POOL_SIZE,
+                                                       sizeof( Command_t * ),
+                                                       staticQueueStorageArea,
+                                                       &staticQueueStructure );
+        configASSERT( commandStructQueue.queue );
+
+        /* Populate the queue. */
+        for( i = 0; i < MQTT_COMMAND_CONTEXTS_POOL_SIZE; i++ )
+        {
+            /* Store the address as a variable. */
+            pCommand = &commandStructurePool[ i ];
+            /* Send the pointer to the queue. */
+            commandAdded = Agent_MessageSend( &commandStructQueue, &pCommand, 0U );
+            configASSERT( commandAdded );
+        }
+
+        initStatus = QUEUE_INITIALIZED;
+    }
+}
+
+/*-----------------------------------------------------------*/
+
+Command_t * Agent_GetCommand( uint32_t blockTimeMs )
+{
+    Command_t * structToUse = NULL;
+    size_t i;
+    bool structRetrieved = false;
+
+    /* Check queue has been created. */
+    if( initStatus == QUEUE_INITIALIZED )
+    {
+        /* Retrieve a struct from the queue. */
+        structRetrieved = Agent_MessageReceive( &commandStructQueue, &( structToUse ), blockTimeMs );
+
+        if( !structRetrieved )
+        {
+            LogError( ( "No command structure available." ) );
+        }
+    }
+
+    return structToUse;
+}
+
+/*-----------------------------------------------------------*/
+
+bool Agent_ReleaseCommand( Command_t * pCommandToRelease )
+{
+    size_t i;
+    bool structReturned = false;
+
+    /* See if the structure being returned is actually from the pool. */
+    if( ( pCommandToRelease >= commandStructurePool ) &&
+        ( pCommandToRelease < ( commandStructurePool + MQTT_COMMAND_CONTEXTS_POOL_SIZE ) ) )
```

Comment:
Don't we have to make sure the Queue is initialized here?

Reply:
Added an assert